### ISSUE TITLE
RFE: container-native rpmdb format

### ISSUE THREAD

**Issue by cgwalters** [rpm-software-management/rpm#2005]
The way OCI/Docker containers work is a [series of layers](https://github.com/opencontainers/image-spec/blob/main/layer.md).  In overlayfs, modified files are "copied up".

There are two issues:

 - rpms have a lot of metadata, so the rpmdb can be of nontrivial size
 - The current main rpmdb format (sqlite) being a single big file means that it gets duplicated in each derived image

In the common case of something like e.g.

```
FROM registry.fedoraproject.org/fedora:35
RUN dnf -y install cowsay && dnf clean all
```

The resulting tar layer from the `RUN` command has an *entirely duplicated* copy of the rpm database, just with `cowsay` and its dependencies added.

It's quite common for builds like this to actually form the base image for further images - and this duplication gets compounded.

A strawman proposal here is something like `/usr/lib/sysimage/rpmdb.d` with something simple like zstd-compressed JSON files storing data instead.  The files could be named something simple like `0000.json.zstd` and then later changes which add packages add just a new `0001.json.zstd` file or so.  Or, we could keep sqlite and union those; I don't have a really strong opinion.  (Well, not really "union" literally but more "merge", since we should support removing or upgrading packages from prior layers)

**Comment by voxik** [comment]
This is either related or duplicate of #1885.

**Comment by cgwalters** [comment]
You're right, this overlaps a lot with previous threads, however, this one is much more about OCI/Docker containers than the previous ones.

**Comment by DemiMarie** [comment]
There are two approaches to fix this:

1. Use a storage engine that provides block-level copy-on-write, rather than file-level copy-on-write.  BTRFS, ZFS, and device-mapper satisfy this requirement, as does overlay2 on a filesystem supporting reflinks.  This _does not_ help image layer sizes, however.
2. Use a one-file-per-entry approach.

**Comment by p5** [comment]
Created an (AI-assisted) PoC of this, and appears to be functional after limited testing.

Each installed RPM is represented by a numbered file in `/var/lib/rpm/jsonzstd`, such as `00000001.json.zstd`.  Each file contains the RPM header data as a B64 encoded string, and is rebuilt in-memory when required.  

```json
{
  "version": 1,
  "hdrNum": 1,
  "blob": "<base64-encoded RPM header blob>"
}
```

In the same directory lives a "manifest" file which points to each individual package file and stores metadata about the next number in the sequence.  This manifest still has the same drawback of the sqlite db, where it will be stored across multiple layers, but is considerably smaller.  This could be adapted to no longer require the manifest file, though would require some other method for calculating the next available file index.

```json
{
  "version": 1,
  "nextHdrNum": 42,
  "packages": [1, 2, 5, 7, 12]
}
```

RPM is able to rebuild the jsonzstd database from a pre-existing sqlite database, and packages are installable via `dnf` and query-able via the updated `rpm`.

---

The implementation can be found in my forked repo - https://github.com/p5/rpm-software-management-rpm/tree/jsonzstd-rpmdb-backend - and I am currently attempting to publish a build in a COPR to ease testing.

```
$ podman build -t rpm-jsonzstd -f tests/Dockerfile.jsonzstd .
$ podman run --rm -it -e RPM_CONFIGDIR=/usr/local/lib/rpm rpm-jsonzstd /bin/bash
# rpm --rebuilddb
# dnf install -y jq
```

Demo (access controlled) - https://drive.google.com/file/d/1GvSAFEHuI-2TiVGA93I_f3_r1f_IRFv-/view?usp=drive_link

---

I am fully aware of this project's stance on AI generated code and am in no way claiming this implementation is production-ready, free of bugs or even fully-functional.  This comment/fork serves as a proof-of-concept for how this could look, and to demonstrate the usefulness of a backend more optimised for container images.

**Comment by cgwalters** [comment]
> Each installed RPM is represented by a numbered file in /var/lib/rpm/jsonzstd,

This should be in the default `%_dbpath` which shouldn't be `/var` except on RHEL <= 9.

> In the same directory lives a "manifest" file which points to each individual package file and stores metadata about the next number in the sequence. 

Is the sequence defined by package installation order for a fresh image? And then subsequent ones use N+1 starting from the last? 

> This comment/fork serves as a proof-of-concept for how this could look, and to demonstrate the usefulness of a backend more optimised for container images.

Yes, thanks for doing that! 

This issue is still relevant; I forgot to do an explicit cross reference but e.g. https://github.com/rpm-software-management/dnf5/issues/1731 would be a *lot* more elegant with this, and that's a not-OCI-container use case.

**Comment by p5** [comment]
> > Each installed RPM is represented by a numbered file in /var/lib/rpm/jsonzstd,
> 
> This should be in the default %_dbpath which shouldn't be /var except on RHEL <= 9.

I will make that change!  Though please note that I have encountered different problems with the implementation (nothing major, just due to lack of understanding of the codebase/existing backend) which prevented signature verification to work.  I do want to pick this back up and make something fully functional but I don't believe I can contribute back here due to the AI policy and my limited C knowledge.

> > In the same directory lives a "manifest" file which points to each individual package file and stores metadata about the next number in the sequence.
> 
> Is the sequence defined by package installation order for a fresh image? And then subsequent ones use N+1 starting from the last?

Exactly.  Similar to the current sqlite backend, the order of the database (in this case the files) is determined by the order the packages were installed.  Thought this made the most sense given the scriptlets being invoked in different orders could result in filesystem differences.

The manifest file could be done away with, but I liked the thought of leaving a blank number where a package is removed which requires storing the next available value in a file.  Potentially allowing for hotplugging packages or similar.

With regards to the sysext use-case, definitely!  I had not considered that use-case, I was more thinking about derived container images.
